gpu/gpu_info_cuda.c

```diff
@@ -82,121 +86,283 @@ void cuda_init(char *cuda_lib_path, cuda_init_resp_t *resp) {
-
-  if (h.handle == NULL) {
-    resp->err = strdup("nvml handle sn't initialized");
-    return;
-  }
 
-  ret = (*h.nvmlDeviceGetCount_v2)(&resp->count);
+  ret = (*h.nvmlDeviceGetCount_v2)(&numDevices);
   if (ret != NVML_SUCCESS) {
     snprintf(buf, buflen, "unable to get device count: %d", ret);
-    resp->err = strdup(buf);
-    return;
+    *err = strdup(buf);
+    return 0;
   }
 
-  resp->total = 0;
-  resp->free = 0;
-  for (i = 0; i < resp->count; i++) {
+  deviceMap = new_device_map(numDevices);
+
+  for(int i = 0;i < numDevices;i++) {
+    unsigned int currentMode = 0,pendingMode;
+
     ret = (*h.nvmlDeviceGetHandleByIndex)(i, &device);
     if (ret != NVML_SUCCESS) {
       snprintf(buf, buflen, "unable to get device handle %d: %d", i, ret);
-      resp->err = strdup(buf);
-      return;
+      *err = strdup(buf);
+      free_device_map(deviceMap);
+      return 0;
     }
 
-    ret = (*h.nvmlDeviceGetMemoryInfo)(device, &memInfo);
-    if (ret != NVML_SUCCESS) {
-      snprintf(buf, buflen, "device memory info lookup failure %d: %d", i, ret);
-      resp->err = strdup(buf);
-      return;
+    ret = (*h.nvmlDeviceGetMigMode)(device,&currentMode,&pendingMode);
+    if (ret != NVML_SUCCESS && ret != NVML_ERROR_NOT_SUPPORTED) {
+      snprintf(buf, buflen, "unable to get MIG mode for device %d: %d", i, ret);
+      *err = strdup(buf);
+      free_device_map(deviceMap);
+      return 0;
     }
-    if (h.verbose) {
-      nvmlBrandType_t brand = 0;
-      // When in verbose mode, report more information about
-      // the card we discover, but don't fail on error
-      ret = (*h.nvmlDeviceGetName)(device, buf, buflen);
-      if (ret != RSMI_STATUS_SUCCESS) {
-        LOG(h.verbose, "nvmlDeviceGetName failed: %d\n", ret);
-      } else {
-        LOG(h.verbose, "[%d] CUDA device name: %s\n", i, buf);
-      }
-      ret = (*h.nvmlDeviceGetBoardPartNumber)(device, buf, buflen);
-      if (ret != RSMI_STATUS_SUCCESS) {
-        LOG(h.verbose, "nvmlDeviceGetBoardPartNumber failed: %d\n", ret);
-      } else {
-        LOG(h.verbose, "[%d] CUDA part number: %s\n", i, buf);
-      }
-      ret = (*h.nvmlDeviceGetSerial)(device, buf, buflen);
-      if (ret != RSMI_STATUS_SUCCESS) {
-        LOG(h.verbose, "nvmlDeviceGetSerial failed: %d\n", ret);
-      } else {
-        LOG(h.verbose, "[%d] CUDA S/N: %s\n", i, buf);
-      }
-      ret = (*h.nvmlDeviceGetVbiosVersion)(device, buf, buflen);
-      if (ret != RSMI_STATUS_SUCCESS) {
-        LOG(h.verbose, "nvmlDeviceGetVbiosVersion failed: %d\n", ret);
-      } else {
-        LOG(h.verbose, "[%d] CUDA vbios version: %s\n", i, buf);
-      }
-      ret = (*h.nvmlDeviceGetBrand)(device, &brand);
-      if (ret != RSMI_STATUS_SUCCESS) {
-        LOG(h.verbose, "nvmlDeviceGetBrand failed: %d\n", ret);
-      } else {
-        LOG(h.verbose, "[%d] CUDA brand: %d\n", i, brand);
+    else if(h.verbose && ret != NVML_ERROR_NOT_SUPPORTED) LOG(h.verbose,"MIG Mode is %d\n", currentMode);
```

Comment:
nit - it would be nice to reshuffle this so it is reported in the verbose set of attributes for the GPU, device name, etc, and MIG status, and maybe count of virtual devices if that is in fact dynamic.